### PR TITLE
MI-260: Create API endpoint for login with LTI and validate LTI Launch fields

### PR DESCRIPTION
**Pull request by chetanyakan**
#### Summary
- Create an API endpoint to Login with LTI which takes a POST request with LTI launch fields.
- Iterate over LTI config to find consumer secret using consumer key in the request
- Validate consumer key by validating OAuth Signature of the LTI launch fields.

#### Ticket Link
https://brightscout.atlassian.net/browse/MI-257
https://brightscout.atlassian.net/browse/MI-260
https://brightscout.atlassian.net/browse/MI-261

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: app/lti.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
```
Comment (harshilsharma63) [inline app/lti.go]: I don't think this code belongs to the app layer of Mattermost. It's more of a utility.

File: model/config.go [hunk 2]
```diff
@@ -633,6 +633,29 @@ func (s *AnalyticsSettings) SetDefaults() {
 	}
 }
 
+type LMSOAuthSettings struct {
```
Comment (harshilsharma63) [inline model/config.go]: Refactor the types as discussed.

File: web/lti.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/mattermost/mattermost-server/app"
+	"github.com/mattermost/mattermost-server/mlog"
+)
+
+func (w *Web) InitLti() {
+	w.MainRouter.Handle("/login/lti", w.NewHandler(loginWithLti)).Methods("POST")
+}
+
+func loginWithLti(c *Context, w http.ResponseWriter, r *http.Request) {
+
```
Comment (harshilsharma63) [inline web/lti.go:22]: Please remove the empty line.

**Review by harshilsharma63 — CHANGES_REQUESTED** [review]

File: web/lti.go (hunk 3 above)
Comment (chetanyakan) [reply web/lti.go:22]: Done

File: model/config.go (hunk 2 above)
Comment (chetanyakan) [reply model/config.go]: Done

File: config/config-dev.json [hunk 4]
```diff
@@ -326,6 +326,58 @@
         "LoginButtonBorderColor": "",
         "LoginButtonTextColor": ""
     },
+    "LtiSettings": {
```
Comment (harshilsharma63) [inline config/config-dev.json]: This should be `LTISettings` (capital LTI).

File: model/config.go [hunk 5]
```diff
@@ -633,6 +637,62 @@ func (s *AnalyticsSettings) SetDefaults() {
...
+	NameProperty string
+	IDProperty   string
+}
+
+type EdxUserChannelsSettings struct {
+	Type        string
+	ChannelList []EdxChannel
+}
+
+type EdxLMSSettings struct {
```
Comment (harshilsharma63) [inline model/config.go]: Formatting seems to be a bit off over here. Can you run `gofmt` on the file?

File: model/config.go [hunk 6]
```diff
@@ -633,6 +637,62 @@ func (s *AnalyticsSettings) SetDefaults() {
 	}
 }
 
+type LMSOAuthSettings struct {
```
Comment (harshilsharma63) [inline model/config.go]: Let's move all this logic to `model/lti.go` to avoid conflicts with Mattermost.

File: model/config.go [hunk 7]
```diff
@@ -633,6 +637,62 @@ func (s *AnalyticsSettings) SetDefaults() {
...
+	IDProperty   string
+}
+
+type EdxUserChannelsSettings struct {
+	Type        string
+	ChannelList []EdxChannel
+}
+
+type EdxLMSSettings struct {
+	Name  string
```
Comment (harshilsharma63) [inline model/config.go]: This should extend `LMSSettings`.

File: web/lti.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+		return
+	}
+
+	// Validate request
+	lmss := c.App.Config().LTISettings.GetKnownLMSs()
+	ltiConsumerKey := r.FormValue("oauth_consumer_key")
+	var ltiConsumerSecret string
+
+	for _, val := range lmss {
+		if lms, ok := val.(model.EdxLMSSettings); ok {
```
Comment (harshilsharma63) [inline web/lti.go:30]: We should cast this to `LMSSettings` and not any of the sub-types.

File: web/lti.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+			if lms.OAuth.ConsumerKey == ltiConsumerKey {
+				ltiConsumerSecret = lms.OAuth.ConsumerSecret
+				break
+			}
+		}
+	}
+
```
Comment (harshilsharma63) [inline web/lti.go:37]: Can we check over here whether we found a consumer secret or not? If not found we can add a log statement about that.

**Review by harshilsharma63 — CHANGES_REQUESTED** [review]

File: web/lti.go (hunk 9 above)
Comment (chetanyakan) [reply web/lti.go:37]: Done

File: model/config.go (hunk 5 above)
Comment (chetanyakan) [reply model/config.go]: Done

File: model/config.go (hunk 6 above)
Comment (chetanyakan) [reply model/config.go]: Done

File: model/config.go (hunk 7 above)
Comment (chetanyakan) [reply model/config.go]: Done

File: web/lti.go (hunk 8 above)
Comment (chetanyakan) [reply web/lti.go:30]: As discussed, added a todo to handle that later

File: web/lti.go (hunk 9 above)
Comment (chetanyakan) [reply web/lti.go:37]: Done

File: config/config-dev.json (hunk 4 above)
Comment (chetanyakan) [reply config/config-dev.json]: Done

**Review by harshilsharma63 — APPROVED** [review]

**Pull request merged**